atomistics/calculators/wrapper.py

```diff
@@ -0,0 +1,81 @@
+"""
+A wrapper for mapping between functions that evaluate a single structure to those
+that evaluate a task dictionary.
+"""
+
+from __future__ import annotations
+
+from typing import NewType, Union, Any, TYPE_CHECKING
+
+from atomistics.calculators.interface import TaskEnum, TaskOutputEnum
+
+if TYPE_CHECKING:
+    from ase import Atoms
+    from atomistics.calculators.interface import (
+        TaskName,
+        TaskSpec,
+        TaskDict,
+        TaskResults,
+        ResultsDict,
+        SimpleEvaluator,
+    )
+
+
+def _convert_task_dict(old_task_dict: dict[TaskName, dict[str, Atoms]]) -> TaskDict:
+    """
+    Converts the existing task dictionaries of the format
+    `{result_type_string: {structure_label_string: structure, ...}, ...}`
+    to the new format
+    `{structure_label_string: (structure, [result_type_string, ...]), ...}`.
+
+    Can be removed if/when the rest of the codebase passing in these task
+    dictionaries gets updated to the new format.
+    """
+    task_dict = {}
+    for method_name, subdict in old_task_dict.items():
+        for label, structure in subdict.items():
+            try:
+                task_dict[label][1].append(method_name)
+            except KeyError:
+                task_dict[label] = (structure, [method_name])
+    return task_dict
+
+
+def as_task_dict_evaluator(
+    calculate: SimpleEvaluator,
+) -> callable[[dict[TaskName, dict[str, Atoms]], ...], ResultsDict]:
+    """
+    Takes a callable that acts on a single structure and a (string) list of tasks to
+    and maps it to a function that operates on a task-list dictionary of structures,
+    structure labels, and the same task list strings. Similarly, maps the output from a
+    single dictionary of task-name-related-output-labels to a nested dictionary using
+    both the output labels and the structure labels.
+
+    Args:
+        calculate [callable]: The function that interprets structures into physical
+            properties.
+
+    Returns:
+        callable: The function operating on a different space.
+    """
+
+    def evaluate_with_calculator(
+        task_dict: dict[TaskName, dict[str, Atoms]],
+        # TODO: Make workflows pass task dicts: dict[str, TaskSpec] ~ TaskDict,
+        *calculate_args,
+        **calculate_kwargs,
+    ) -> ResultsDict:
+        task_dict = _convert_task_dict(task_dict)
+        results_dict = {}
+        for label, (structure, tasks) in task_dict.items():
+            tasks = [TaskEnum(t) for t in tasks]
+            output = calculate(structure, tasks, *calculate_args, **calculate_kwargs)
+            for task_name in tasks:
+                result_name = TaskOutputEnum(task_name).name
```

Comment:
Following the "Explicit vs. implicit" - I feel this essential loop over the tasks is now hidden in the decorator, which makes the code harder to understand.

Reply:
The decorator "explicitly" says what it is doing. This way developers only need to write a function that handles how to run a single calculation. 

In your previous implementation you were running two separate loops each doing the expensive part of running the calculations. By hiding the loop away we prevent future users from making similar mistakes.

Since the loop is fundamental and going to happen in _every_ evaluator, it should be abstracted somehow. There are no classes, and everything is still completely functional. This meets all the criteria discussed in the meeting and, IMO, makes writing new evaluators easier.

Reply:
A different way to implement this would be to already do the conversion on the `workflow` side and basically return the new kind of task list from the workflow.

Reply:
> A different way to implement this would be to already do the conversion on the `workflow` side and basically return the new kind of task list from the workflow.

My understanding of the current paradigm is this: `Workflows` should generate a dictionary of structures and tasks in their "generate" stage, the user should evaluate the requested properties ("tasks") and provide a similarly-formatted dictionary, then the workflow should do analysis on that dictionary.

If we move the loop stuff inside the `workflow`, this paradigm is violated. The current implementation makes it so the user doesn't _need_ to implement the loops, but they _may_ still implement the loops if they want to (e.g. there is some technical edge case where it's more efficient to do something different than we do in the decorator).

So what's here is an optional tool that makes it easier to obey the existing paradigm, but moving it inside the workflow sounds like a paradigm shift. I am not totally closed to that paradigm shift, but what I merged here is (a) still useful if we don't make that shift, and (b) a good intermediate step if we do make that shift, since the looping is already encapsulated somewhere. 

I'm not closed to such a paradigm shift, but paradigm shifts were intentionally avoided here. All this gave is a shortcut designed to make it easier to implement new calculators.

> I dislike the new interface as a new developer who wants to contribute both a new `workflow` and a new `calculator` has to learn two interfaces.

I don't follow. Both `workflow` and `calculator` developers need to know the magic codes for tasks -- i.e. the mapping between strings and physical properties -- currently "(calc_)energy" and "(calc_)forces". A `workflow` developer additionally needs to understand the structure of task dicts. In the previous implementation,`calculator`-developers-`workflow`-users also _needed_ to understand the structure of task dicts, now, with the use of the `@as_task_dict_evaluator` this knowledge is _optional_ and they only need to know about the magic strings. So nothing new is required of `workflow` developers, and `calculator` developers keep all their freedom and power while gaining a lower barrier to entry.